design/design-doc-composition-validating-webhook.md

```diff
@@ -0,0 +1,391 @@
+       source object, if defined for the type of patch and required by the
+       `PatchPolicy` defined.
+    2. have a valid path as a `toFieldPath` according to the schema of the
+       destination object, if defined for the type of patch.
+    3. have correctly configured types, e.g. if `fromFieldPath` is a string,
+       `toFieldPath` is an integer, the patch should be rejected, but not if
+       there is a mapping correctly setting it.
+3. the result of all the `patches` being applied to the `base` should be valid
+   according to the Managed Resource schema.
+
+Compositions are also defining the following fields that we should take into
+consideration while validating:
+
+- `patchSets`: a list of named patches that can be referenced by resources.
+  Depending on the type of patch it could be impossible to validate them
+  independently from the resource they would be used on. We could think of only
+  validating them once they are dereferenced and associated to a specific
+  resource.
+- `functions`: a list of Composition Functions that can be used to generate or
+  modify the Composed Resources. Given that it’s an `alpha` feature, we can
+  think of not covering this in the first implementation, however we should
+  keep in mind that having at least a function defined would invalidate the
+  assumption that rendered Managed Resources should be valid according to the
+  schema. Moreover, taking them into consideration during the validation phase
+  would mean adding another external dependency, which could both reduce
+  reliability and add latency.
+- `environment`: contains a list of `EnvironmentConfigs` that can be referenced
+  by specific types of patches and a list of`EnvironmentPatches` . Also an
+  `alpha` feature, similar implications to what already said for `functions`,
+  would require additional information and can invalidate the assumption that
+  we render valid Composed Resources if defined.
+- `writeConnectionSecretsToNamespace` : the name of the namespace the `Secret`
+  containing the connection details should be created into. We could check
+  whether it exists already, but that would be against the behaviour Kubernetes
+  users are used to, e.g. a `Deployment` mounting a `Secret` which does not
+  exist yet is considered valid but remains pending until it’s not created.
+- `publishConnectionDetailsWithStoreConfigRef` : a reference by name to a
+  `StoreConfig`, we could check that it exists, but similarly to the namespace
+  it would go against the behaviour Kubernetes users might be expecting.
+
+### Additional Constraints for the ValidatingWebhook use case
+
+A few constraints that makes validating Compositions from a Validating Webhook
+harder w.r.t. the cases mentioned in the Background:
+
+1. The CRDs of the Managed Resources defined as `base` and traversed by
+   `patches` might not yet have been created when the Composition is being
+   validated.
```

Comment:
This makes me wonder whether we should have both a validation webhook and a long-running validation controller.

It seems like a Composition that is valid at creation time (i.e. webhook time) could later become invalid. For example if the CRD for one of the MRs it composes is removed, or (in the case of alpha MRs) there's a breaking schema change.

Such a controller could toggle a `valid` status condition on the `Composition`, preventing it from being selected by XRs.

Reply:
Fair point, indeed given how complex and powerful Compositions are, there are a lot of possible bad scenarios. 

Right now we have no validation at all and errors are reported only once and if a Composition is selected by an XR. Having a validating Webhook would allow us and users to be at least sure that Compositions are valid at deploy time, if all requirements were met. Which paired with a GitOps approach also will let users also know whether Compositions became invalid for any reason.

Obviously later on we might want to address the run time too and we'll have to decide how to address questions like the following ones:
- what happens if the requirements were not met at deploy time, but are met later on and make the Composition invalid? Successive applies of the Composition will fail, but should we be able to notice it and report it in some way?
- what happens if all requirements were met at deploy time and therefore the resource was fully validated and later on one of the requirements is not satisfied anymore? Successive applies will still succeed, but we'll have less confidence and depending on whether there are already XRs using it or not how should these be handled?
- what if it was fully validated at deploy time and, as you are suggesting, one of the requirements changes and makes the Composition invalid? Successive applies of the Composition will fail, but what should happen for existing and new XRs using it? How are users going to be notified?

Ideally we would want to only accept fully validated Compositions, always know whether a Composition is valid or not, prevent any action that would invalidate a valid Composition and report any error on the Composition itself as soon as possible.

However, given that currently users expect not to have any validation at all, we can not be too aggressive on it, so I'd suggest we start from the Webhook and iterate from there one step at a time.

Reply:
> However, given that currently users expect not to have any validation at all, we can not be too aggressive on it, so I'd suggest we start from the Webhook and iterate from there one step at a time.

Absolutely - I should have been explicit I didn't mean to bring any of this into scope for this current design.